taar/context.py

```diff
@@ -17,8 +19,78 @@
 chain.
 """
 
-# Clobber the Context name to prevent messy name collisions
-from srgutil.context import default_context as _default_context
+
+class InvalidInterface(Exception):
+    """Raise this when impl() fails to export an implementation"""
+    pass
+
+
+class Context:
+    def __init__(self, delegate=None):
+        if delegate is None:
+            delegate = {}
+
+        self._local_dict = {}
+        self._delegate = delegate
+
+    def __contains__(self, key):
+        try:
+            self[key]
+            return True
+        except KeyError:
+            return False
+
+    def __getitem__(self, key):
+        # This is a little tricky, we want to lookup items in our
+        # local namespace before we hit the delegate
+        try:
+            result = self._local_dict[key]
+        except KeyError:
+            result = self._delegate[key]
+        return result
+
+    def get(self, key, default):
+        try:
+            result = self[key]
+        except KeyError:
+            result = default
+        return result
+
+    def __setitem__(self, key, value):
+        self._local_dict[key] = value
+
+    def __delitem__(self, key):
+        del self._local_dict[key]
+
+    def wrap(self, ctx):
+        ctx_child = ctx.child()
+        this_child = self.child()
+        this_child._delegate = ctx_child
+        return this_child
+
+    def child(self):
+        """ In general, you should call this immediately in any
+        constructor that receives a context """
+
+        return Context(self)
+
+    def impl(self, iface):
+        instance = self._local_dict[iface]
+        if not isinstance(instance, iface):
+            raise InvalidInterface("Instance [%s] doesn't implement requested interface.")
+        return instance
+
+
+def _default_context():
+    ctx = Context()
+    from taar.logs import Logging
+    from taar.logs import IMozLogging
+    from decouple import config
+
+    logger = Logging(ctx)
+    logger.set_log_level(config("PYTHON_LOG_LEVEL", "INFO"))
```

Comment:
I think we should manage the `PYTHON_LOG_LEVEL` in https://github.com/mozilla/taar/blob/master/taar/settings.py instead of calling decouple.config directly here and import taar.settings to pull in the value. This will allow us to maintain all the configurable options in one area.

Reply:
agree, but the reason I put it here is that `settings.py` is used only in taar service but `context.py` is also used from Spark jobs as a part of taar package, so there shouldn't be a dependency on `settings` from those jobs. But you're right it will be hard to find it here, let me see what I can do about it